pkg/ipam/poolallocator/testing/fake_client.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testing
+
+import (
+	"sync"
+
+	"github.com/google/uuid"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	k8stesting "k8s.io/client-go/testing"
+
+	crdv1a2 "antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	fakeversioned "antrea.io/antrea/pkg/client/clientset/versioned/fake"
+)
+
+// Simple client is not sufficient for pool allocator testing,
+// since pool allocator relies on both crd client and pool informer
+// to work in sync. This client extension mimics the real client in
+// conflict handling functionality - pool update will return conflict
+// error unless ResourceVersion for the updated pool reflect the version
+// stored in the client.
+type IPPoolClientset struct {
+	fakeversioned.Clientset
+	// store latest ResourceVersion for given pool
+	poolVersion sync.Map
+	watcher     *watch.RaceFreeFakeWatcher
+}
+
+func (c *IPPoolClientset) InitPool(pool *crdv1a2.IPPool) {
+	pool.ResourceVersion = uuid.New().String()
+	c.poolVersion.Store(pool.Name, pool.ResourceVersion)
+
+	c.watcher.Add(pool)
+}
+
+func NewIPPoolClient() *IPPoolClientset {
+
+	crdClient := &IPPoolClientset{watcher: watch.NewRaceFreeFake()}
+	// map needs to be != nil
+	crdClient.poolVersion.Store("placeholder", nil)
```

Comment:
I don't quite understand this. It doesn't seem nil value is checked anywhere.

Reply:
I think an empty map is not persisted between threads, we need to place a dummy object for the map to initialize. I've added a more detailed comment here.

Reply:
Indeed this didn't make sense to me, so I dug a little and discovered that I forgot to initialize the map :) So this is fixed now and thanks for your comments here.